File: day-01/part-01.hs
import Control.Arrow ((>>>))
import Control.Exception (assert)

solve :: [String] -> Int
solve = foldl processLine (50, 0) >>> snd

processLine (pos, count) line =
    let (dir:rest) = line
        value = read rest :: Int
        delta = if dir == 'L' then -value else value
        newPos = (pos + delta) `mod` 100
        newCount = if newPos == 0 then count + 1 else count
    in (newPos, newCount)

main :: IO ()
main = do
  -- Test with example
  exampleContent <- readFile "day-01/example.txt"
  let exampleResult = solve (lines exampleContent)
  assert (exampleResult == 3) $ return ()

  -- Solve with input
  inputContent <- readFile "day-01/input.txt"
  let result = solve (lines inputContent)
  print result

File: day-01/part-01.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(lines []string) int {
	position := 50
	count := 0

	for _, line := range lines {
		if len(line) == 0 {
			continue
		}

		direction := line[0]
		value, _ := strconv.Atoi(line[1:])

		if direction == 'L' {
			position -= value
		} else {
			position += value
		}

		for position < 0 {
			position += 100
		}
		for position >= 100 {
			position -= 100
		}

		if position == 0 {
			count++
		}
	}

	return count
}

func readLines(filename string) ([]string, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	text := strings.TrimSpace(string(content))
	return strings.Split(text, "\n"), nil
}

func main() {
	// Test with example
	exampleLines, _ := readLines("day-01/example.txt")
	exampleResult := solve(exampleLines)
	if exampleResult != 3 {
		fmt.Printf("Example failed: expected 3, got %d\n", exampleResult)
		os.Exit(1)
	}

	// Solve with input
	inputLines, _ := readLines("day-01/input.txt")
	result := solve(inputLines)
	fmt.Println(result)
}
